File: src/cammon.c
#define _GNU_SOURCE
#include <ctype.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <libavformat/avformat.h>
#include <pthread.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/inotify.h>
#include <sys/select.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <time.h>
#include <unistd.h>

// gcc -Os -o bin/cammon src/cammon.c -lavformat -lavutil -lavcodec -pthread

typedef struct {
  char *maxfile;
  char *prefix;
  char *logfile;
} config;

void *filter(void *input) {
  const char *in_filename = (const char *)input;
  char cmd[1024];
  if (fcntl(stdout->_fileno, F_SETPIPE_SZ, 134217728) == 134217728) {
    int flags = fcntl(stdout->_fileno, F_GETFL, 0);
    if (flags != -1) fcntl(stdout->_fileno, F_SETFL, flags | O_NONBLOCK);
  }
  snprintf(cmd, sizeof(cmd),
           "ffmpeg -vsync 0 -loglevel quiet -nostdin -i %s -f rawvideo "
           "-codec:v copy -codec:a none -bsf:v h264_mp4toannexb -",
           in_filename);
  return system(cmd) ? NULL : input;
}

void *filter_new(void *input) {
  const char *in_filename = (const char *)input;
  AVOutputFormat *ofmt_v = NULL;
  AVFormatContext *ifmt_ctx = NULL, *ofmt_ctx_v = NULL;

  if (fcntl(stdout->_fileno, F_SETPIPE_SZ, 134217728) == 134217728) {
    int flags = fcntl(stdout->_fileno, F_GETFL, 0);
    if (flags != -1) fcntl(stdout->_fileno, F_SETFL, flags | O_NONBLOCK);
  }
  int videoindex = -1;
  av_register_all();
  av_log_set_level(AV_LOG_QUIET);

  // Input
  if (avformat_open_input(&ifmt_ctx, in_filename, 0, 0) < 0) return input;

  // Output
  avformat_alloc_output_context2(&ofmt_ctx_v, NULL, "h264", NULL);
  if (!ofmt_ctx_v) return input;

  ofmt_v = ofmt_ctx_v->oformat;

  for (int i = 0; i < ifmt_ctx->nb_streams; i++) {
    if (ifmt_ctx->streams[i]->codecpar->codec_type == AVMEDIA_TYPE_VIDEO) {
      videoindex = i;
      avformat_new_stream(ofmt_ctx_v, NULL);
      break;
    }
  }

  if (videoindex < 0 || (!(ofmt_v->flags & AVFMT_NOFILE) && avio_open(&ofmt_ctx_v->pb, "/dev/stdout", AVIO_FLAG_WRITE) < 0) ||
      avformat_write_header(ofmt_ctx_v, NULL) < 0)
    return input;

  AVBSFContext *h264bsfc;
  const AVBitStreamFilter *bsf = av_bsf_get_by_name("h264_mp4toannexb");
  if (!bsf || av_bsf_alloc(bsf, &h264bsfc)) return input;

  AVPacket pkt;
  while (av_read_frame(ifmt_ctx, &pkt) >= 0) {
    AVStream *in_stream = ifmt_ctx->streams[pkt.stream_index];

    if (pkt.stream_index == videoindex) {
      avcodec_parameters_copy(h264bsfc->par_in, in_stream->codecpar);
      if (av_bsf_init(h264bsfc) < 0 || av_bsf_send_packet(h264bsfc, &pkt) < 0) {
        av_bsf_free(&h264bsfc);
        av_packet_unref(&pkt);
        return input;
      }

      if (av_bsf_receive_packet(h264bsfc, &pkt) != AVERROR(EAGAIN)) av_interleaved_write_frame(ofmt_ctx_v, &pkt);
    }
    av_packet_unref(&pkt);
  }

  av_bsf_free(&h264bsfc);

  // Write file trailer
  av_write_trailer(ofmt_ctx_v);
  avformat_close_input(&ifmt_ctx);

  /* Close output */
  if (ofmt_ctx_v && !(ofmt_v->flags & AVFMT_NOFILE)) avio_close(ofmt_ctx_v->pb);

  avformat_free_context(ofmt_ctx_v);
  return NULL;
}

#define EVENT_SIZE (sizeof(struct inotify_event))
#define BUF_LEN (1024 * (EVENT_SIZE + 16))

static unsigned long hash(char *str) {
  unsigned long hash = 5381;
  int c;

  while ((c = *str++)) hash = ((hash << 5) + hash) + c;

  return hash;
}

static void printlog(const char *log, const char *format, ...) {
  FILE *f = fopen(log, "a");
  if (f) {
    va_list args;
    char tsbuf[25];
    time_t ts = time(NULL);
    struct tm *t = gmtime(&ts);
    strftime(tsbuf, sizeof(tsbuf), "%Y-%m-%d %H:%M:%S", t);
    fprintf(f, "%s: ", tsbuf);
    va_start(args, format);
    vfprintf(f, format, args);
    va_end(args);
    fclose(f);
  }
}

static int yesterday_today_tomorrow(const char *fmt, size_t len, char *yesterday, char *today, char *tomorrow) {
  time_t ts = time(NULL);
  struct tm *t = gmtime(&ts);
  unsigned long before = hash(today);
  strftime(today, len, fmt, t);
  unsigned long after = hash(today);
  ts -= 24 * 3600;
  t = gmtime(&ts);
  strftime(yesterday, len, fmt, t);
  ts += 48 * 3600;
  t = gmtime(&ts);
  strftime(tomorrow, len, fmt, t);
  struct stat st = { 0 };
  if (stat(tomorrow, &st) == -1) mkdir(tomorrow, 0755);
  return before != after;
}

int *watches = 0;
char **watches_names;
int watches_count;
int watches_size = 128;
int fd = -1;

int find_watch(int wd) {
  for (int i = 0; i < watches_count; i++)
    if (watches[i] == wd) return i;
  return -1;
}

void add_watch(const char *name, const char *logfile) {
  if (!watches) {
    watches = malloc(watches_size * sizeof(int));
    watches_names = malloc(watches_size * sizeof(char *));
    watches_count = 0;
  }

  struct stat statbuf;
  stat(name, &statbuf);
  if (S_ISDIR(statbuf.st_mode)) {
    struct dirent *dent;
    DIR *srcdir = opendir(name);
    while ((dent = readdir(srcdir)) != NULL) {
      struct stat st;
      if (strcmp(dent->d_name, ".") == 0 || strcmp(dent->d_name, "..") == 0) continue;
      if (fstatat(dirfd(srcdir), dent->d_name, &st, 0) < 0) {
        perror(dent->d_name);
        continue;
      }
      if (S_ISDIR(st.st_mode)) {
        char buf[1024];
        snprintf(buf, sizeof(buf), "%s/%s", name, dent->d_name);
        add_watch(buf, logfile);
      }
    }
    closedir(srcdir);
  }
  watches_names[watches_count] = malloc(strlen(name) + 1);
  strcpy(watches_names[watches_count], name);
  watches[watches_count++] = inotify_add_watch(fd, name, IN_CLOSE_WRITE | IN_MOVED_TO);
  printlog(logfile, "Adding watch \"%s\" (%d)\n", name, watches_count);
  if (watches_count == watches_size) {
    watches_size += 128;
    watches = realloc(watches, watches_size);
    char **old_names = watches_names;
    watches_names = malloc(watches_size * sizeof(char *));
    for (int i = 0; i < watches_count; i++) {
      watches_names[i] = malloc(strlen(old_names[i]) + 1);
      strcpy(watches_names[i], old_names[i]);
      free(old_names[i]);
    }
    free(old_names);
  }
}

void rm_watches(const char *prefix, const char *logfile) {
  for (int i = 0; i < watches_count; i++) {
    if (!prefix || !strncmp(prefix, watches_names[i], strlen(prefix))) {
      printlog(logfile, "Removing watch \"%s\"\n", watches[i]);
      inotify_rm_watch(fd, watches[i]);
      watches[i] = -1;
      free(watches_names[i]);
      watches_names[i] = 0;
    }
  }
  int j = 0;
  for (int i = 0; i < watches_count; i++) {
    if (watches[i] >= 0) {
      watches[j] = watches[i];
      watches_names[j] = watches_names[i];
      j++;
    }
  }

  watches_count = j;
  printlog(logfile, "Remaining watches: %d\n", watches_count);
}

static void printhelp(const char *prog) {
  fprintf(stderr, "Usage: %s [options] <dir>\n", prog);
  fprintf(stderr,
          " This program will monitor three directories specified by "
          "<dir> for new files\n");
  fprintf(stderr,
          " and extract the raw h.264 streams from them and output "
          "those to stdout.\n");
  fprintf(stderr,
          " <dir> is a string passed to strftime() to format "
          "yesterday, today and tomorrow.\n");
  fprintf(stderr, " -h:           Print this help and exit.\n");
  fprintf(stderr, " -l <logfile>: Log file,\n");
  fprintf(stderr, " -p <logfile>: Prefix for the files to read,\n");
  fprintf(stderr,
          " -x <maxfile>: File which contains the location of an "
          "hourly \"max\" file used\n");
  fprintf(stderr, "               by metdetect.\n");
  fprintf(stderr, " Example: %s -p full_ /meteor/cam1/%%Y%%m%%d\n", prog);
  fprintf(stderr,
          "         If the current date is 2018-01-01, these "
          "directories will be monitored:\n");
  fprintf(stderr, "         /meteor/cam1/20171231\n");
  fprintf(stderr, "         /meteor/cam1/20180101\n");
  fprintf(stderr, "         /meteor/cam1/20180102\n");
  fprintf(stderr, "         Which directories that get monitored are updated with time.\n");
  fprintf(stderr, "         Only files beginning with full_ will be read.\n");
  fprintf(stderr,
          " The intended use for this program is to provide input for "
          "the metdetect program.\n");
}

int parseopts(config *config, int argc, char **argv) {
  int c;

  while ((c = getopt(argc, argv, "hl:p:x:")) != -1) {
    switch (c) {
      case 'h': printhelp(argv[0]); exit(1);
      case 'l':
        if (config->logfile) free(config->logfile);
        config->logfile = strdup(optarg);
        if (!config->logfile) {
          fprintf(stderr, "Memory error\n");
          return -1;
        }
        break;
      case 'p':
        if (config->prefix) free(config->prefix);
        config->prefix = strdup(optarg);
        if (!config->prefix) {
          fprintf(stderr, "Memory error\n");
          return -1;
        }
        break;
      case 'x':
        if (config->maxfile) free(config->maxfile);
        config->maxfile = strdup(optarg);
        if (!config->maxfile) {
          fprintf(stderr, "Memory error\n");
          return -1;
        }
        break;
      case '?':
        if (isprint(optopt))
          fprintf(stderr, "Unknown option `-%c'.\n", optopt);
        else
          fprintf(stderr, "Unknown option character `\\x%x'.\n", optopt);
        return 1;
      default: return 1;
    }
  }
  return 0;
}

pthread_mutex_t running = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t done = PTHREAD_COND_INITIALIZER;

typedef struct {
  int nfds;
  fd_set *set;
  int retval;
  struct timespec *timeout;
  config *config;
} try_select_data;

void *try_select(void *data) {
  int oldtype;
  try_select_data *d = (try_select_data *)data;
  struct timeval timeout;
  memset(&timeout, 0, sizeof(timeout));
  timeout.tv_sec = 120;

  pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, &oldtype);
  d->retval = select(d->nfds, d->set, NULL, NULL, &timeout);
  if (!d->retval) printlog(d->config->logfile, "Select timed out.\n");

  pthread_cond_signal(&done);
  return NULL;
}

/* note: this is not thread safe as it uses a global condition/mutex */
int do_or_timeout(try_select_data *data) {
  struct timespec abs_time;
  pthread_t tid;
  int err;

  pthread_mutex_lock(&running);

  /* pthread cond_timedwait expects an absolute time to wait until */
  clock_gettime(CLOCK_REALTIME, &abs_time);
  abs_time.tv_sec += data->timeout->tv_sec;
  abs_time.tv_nsec += data->timeout->tv_nsec;

  pthread_create(&tid, NULL, try_select, data);

  err = pthread_cond_timedwait(&done, &running, &abs_time);

  if (err == ETIMEDOUT) {
    data->retval = -1;
    printlog(data->config->logfile, "Select froze.\n");
  }
  if (!err) pthread_mutex_unlock(&running);

  return err;
}

int main(int argc, char **argv) {
  config config;
  config.maxfile = config.logfile = NULL;
  config.prefix = strdup("");

  if (argc < 2) {
    printhelp(argv[0]);
    exit(1);
  }

  if (parseopts(&config, argc, argv)) return 1;

  char yesterday[1024];
  char today[1024];
  char tomorrow[1024];
  char buffer[BUF_LEN];
  char lastname[1024];
  char *dir = argv[argc - 1];

  memset(yesterday, 0, sizeof(yesterday));
  memset(today, 0, sizeof(today));
  memset(tomorrow, 0, sizeof(tomorrow));
  memset(lastname, 0, sizeof(lastname));

  while (1) {
    fd_set set;
    struct timespec timeout;
    try_select_data data;
    memset(&timeout, 0, sizeof(timeout));
    memset(&data, 0, sizeof(data));
    timeout.tv_sec = 150;

    do {
      char old_yesterday[1024];
      strncpy(old_yesterday, yesterday, sizeof(old_yesterday));
      if (data.retval < 0 || fd < 0 || yesterday_today_tomorrow(dir, sizeof(yesterday), yesterday, today, tomorrow)) {
        if (fd >= 0) {
          rm_watches(NULL, config.logfile);
          close(fd);
        }
        fd = inotify_init();
        if (fd < 0) perror("inotify_init");
        yesterday_today_tomorrow(dir, sizeof(yesterday), yesterday, today, tomorrow);
        printlog(config.logfile, "Initialising watches: \"%s\", \"%s\", \"%s\"\n", yesterday, today, tomorrow);
        add_watch(yesterday, config.logfile);
        add_watch(today, config.logfile);
        add_watch(tomorrow, config.logfile);
      }

      FD_ZERO(&set);
      FD_SET(fd, &set);
      data.nfds = fd + 1;
      data.set = &set;
      data.timeout = &timeout;
      data.config = &config;

      printlog(config.logfile, "Waiting for new file(s)\n");
      do_or_timeout(&data);
    } while (data.retval < 1 || !FD_ISSET(fd, &set));

    int length = read(fd, buffer, BUF_LEN);

    if (length < 0) perror("read");

    printlog(config.logfile, "Read %d bytes\n", length);

    struct inotify_event *event;
    for (int i = 0; i < length; i += EVENT_SIZE + event->len) {
      event = (struct inotify_event *)&buffer[i];
      if (event->len) {
        char name[1024];
        int index = find_watch(event->wd);
        if (index < 0) continue;
        snprintf(name, sizeof(name), "%s/%s", watches_names[index], event->name);
        struct stat statbuf;
        stat(name, &statbuf);
        if (S_ISDIR(statbuf.st_mode)) {
          printlog(config.logfile, "New dir (%d): %s\n", event->mask, name);
          add_watch(name, config.logfile);
        } else {
          printlog(config.logfile, "New file: %s\n", name);
          if ((event->mask & (IN_MOVED_TO | IN_CLOSE_WRITE)) && strncmp(lastname, name, sizeof(lastname)) &&
              !strncmp(config.prefix, event->name, strlen(config.prefix))) {
            strncpy(lastname, name, sizeof(lastname));
            if (config.maxfile) {
              char mf[1024];
              snprintf(mf, sizeof(mf), "%s/max.jpg", watches_names[index]);
              if (access(mf, F_OK) == -1) {
                FILE *f = fopen(config.maxfile, "w");
                if (f) {
                  fprintf(f, "%s\n", mf);
                  fclose(f);
                }
              }
            }
            printlog(config.logfile, "Filtering %s\n", name);

            pthread_t thread;
            struct timespec ts;

            if (clock_gettime(CLOCK_REALTIME, &ts) == -1 || pthread_create(&thread, NULL, filter_new, name)) {
              printlog(config.logfile, "Error creating thread for filter\n");
              filter_new(name);
            } else {
              int s;
              ts.tv_sec += 180;  // Three minute timeout
              if (pthread_timedjoin_np(thread, NULL, &ts)) {
                printlog(config.logfile, "Filter timed out.\n");
                if (pthread_cancel(thread)) {
                  printlog(config.logfile, "Failed to cancel filter.\n");
                  pthread_join(thread, NULL);
                }
              }
            }
            printlog(config.logfile, "Filtered %s\n", name);
          }
        }
      }
    }
  }

  rm_watches(NULL, config.logfile);
  close(fd);

  free(watches);
  free(watches_names);

  if (config.maxfile) free(config.maxfile);
  if (config.logfile) free(config.logfile);

  return 0;
}

File: bin/cammon.py
#!/usr/bin/env python3

# Usage: cammon.py [-l <logfile>] [-d <format>] [-x <maxfile>] [-b <minutes> ] [ -s <statefile> ] <dir>
# This program monitors directories for new files given by <format>.
# Raw H.264 will be extracted from these and output to stdout for reading by metdetect.

import inotify.adapters
import datetime
import argparse
import os
import pathlib
import time
import glob
import logging
import errno
import signal
import shutil

from ffmpy import FFmpeg

parser = argparse.ArgumentParser(description='Monitor a directory for new files and extract H.264.')
parser.add_argument('-l', '--logfile', dest='logfile', help='logfile')
parser.add_argument('-d', '--format', dest='format', help='file format (default: %Y%m%d/%H/full_%M)', default='%Y%m%d/%H/full_%M')
parser.add_argument('-x', '--maxfile', dest='maxfile', help='file which contains the location of an hourly max file used by metdetect')
parser.add_argument('-b', '--backlog', dest='backlog', help='number of minutes of backlog to look for', default=1)
parser.add_argument('-s', '--statefile', dest='statefile', help='save the most recently processed file to this file')
parser.add_argument(action='store', dest='dir', help='directory to be monitored recursively')
args = parser.parse_args()

logging.basicConfig(format='%(asctime)s: %(message)s', datefmt='%Y-%m-%d %H:%M:%S', filename=args.logfile, level=logging.INFO)

class TimeoutError(Exception):
#     logging.warn('Error processing ' + cur + ": " + e)
#     sys.exit(-1)
    pass

class timeout:
    def __init__(self, seconds=60, error_message='Timeout'):
        self.seconds = seconds
        self.error_message = error_message
    def handle_timeout(self, signum, frame):
        raise TimeoutError(self.error_message)
    def __enter__(self):
        signal.signal(signal.SIGALRM, self.handle_timeout)
        signal.alarm(self.seconds)
    def __exit__(self, type, value, traceback):
        signal.alarm(0)

if not args.dir.endswith('/'):
    args.dir = args.dir + '/'

args.backlog = max(1, int(args.backlog))
i = inotify.adapters.Inotify()
events = i.event_gen()

def add_watches(dirs):
    for dir in dirs:
        logging.info('Add watch: ' + args.dir + dir)
        pathlib.Path(args.dir + dir).mkdir(parents=True, exist_ok=True)
        i.add_watch(args.dir + dir, mask=8+128)

def remove_watches(dirs):
    for dir in dirs:
        logging.info('Remove watch: ' + args.dir + dir)
        i.remove_watch(args.dir + dir)

# Parse timestamp in file and ignore unmatched remainder
def get_time(file):
    return time.strptime(file[:len(datetime.datetime.now().strftime(args.dir + args.format))], args.dir + args.format)
    
def rewind(eventdir, minutes):
    result = []
    for m in reversed(range(0, max(1, minutes))):
        try:
            t = datetime.datetime.fromtimestamp(time.mktime(get_time(eventfile))) - datetime.timedelta(minutes=m)
            result.extend(glob.glob(args.dir + t.strftime(args.format) + '*'))
        except:
            pass
    return result

watches = prev_watches = set()
thishour = nexthour = prevhour = lasteventfile = ''
tophour = datetime.datetime.fromtimestamp(0, datetime.UTC)

import sys
for event in events:
    now = datetime.datetime.now();
    if args.maxfile != None and os.path.dirname(now.strftime(args.format)) != os.path.dirname(tophour.strftime(args.format)):
        tophour = now;
        try:
            # Update the file atomically
            f = open(args.maxfile + '.tmp', 'w')
            f.write(args.dir + os.path.dirname(now.strftime(args.format)) + '/max.jpg\n')
            f.flush()
            f.close()
            os.rename(args.maxfile + '.tmp', args.maxfile)
        except:
            logging.warn('Could not write maxfile')
        
    now = now + datetime.timedelta(minutes=30)
    if os.path.dirname(now.strftime(args.format)) != thishour:
        thishour = os.path.dirname(now.strftime(args.format))
        nexthour = os.path.dirname((now + datetime.timedelta(hours=1)).strftime(args.format))
        prevhour = os.path.dirname((now - datetime.timedelta(hours=1)).strftime(args.format))
        watches = { thishour, nexthour, prevhour }
        remove_watches(prev_watches - watches)
        add_watches(watches - prev_watches)
        prev_watches = watches

    if event == None:
        continue
    (_, _, eventdir, eventfile) = event
    eventfile = eventdir + '/' + eventfile
    file = args.dir + args.format
    try:
        t = get_time(eventfile)
    except:
        t = None

    if t == None or eventfile == lasteventfile:
        continue

    prev = eventfile
    try:
        with open(args.statefile, 'r') as f:
            prev = f.read().strip()
    except:
        if args.statefile != None:
            logging.warn('Could not read statefile')

    lasteventfile = eventfile
    try:
        prevtime = get_time(prev)
    except:
        prevtime = t
    
    files = rewind(eventdir, min(args.backlog, int((time.mktime(t) - time.mktime(prevtime)) / 60)))
    logging.info(files)
    
    for cur in files:
        #sd = cur.replace('full', 'mini')
        jpg_hd = prev.replace('mini', 'full').replace('mp4', 'jpg')
        try:
            shutil.copyfile(jpg_hd, jpg_hd + '.tmp')
            os.rename(jpg_hd + '.tmp', args.dir + '/snapshot.jpg')
        except:
            pass
        #if sd != cur and os.path.isfile(sd):
        #    cur = sd
        ff = FFmpeg(inputs={cur: None}, outputs={'pipe:1': '-nostdin -vsync 0 -timeout 3 -loglevel quiet -f rawvideo -c:v copy -c:a none -bsf:v h264_mp4toannexb'})
        logging.info('Running: ' + ff.cmd)
        try:
            with timeout(seconds=120):
                ff.run()

            logging.info('Finished processing ' + cur + ': ' + str(ff.process.communicate()))
            ff.process.kill()
            if args.statefile != None:
                try:
                    f = open(args.statefile, 'w')
                    f.write(cur + '\n')
                    f.close()
                except:
                    logging.warn('Could not write statefile')

        except Exception as e:
            logging.warning('Error processing ' + cur + ": " + str(e))
